include/nlohmann/json.hpp

```diff
@@ -3693,11 +3704,38 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
         // operator[] only works for objects
         if (JSON_HEDLEY_LIKELY(is_object()))
         {
-            return set_parent(m_value.object->operator[](key));
+            return set_parent(m_value.object->operator[](std::forward<KeyT>(key)));
+        }
+
+        JSON_THROW(type_error::create(305, "cannot use operator[] with a string argument with " + std::string(type_name()), *this));
+    }
+
+#if defined(JSON_HAS_CPP_17)
+    /// @copydoc operator[](KeyT&&)
+    reference operator[](const std::string_view& key)
+    {
+        // implicitly convert null value to an empty object
+        if (is_null())
+        {
+            m_type = value_t::object;
+            m_value.object = create<object_t>();
+            assert_invariant();
+        }
+
+        // operator[] only works for objects
+        if (JSON_HEDLEY_LIKELY(is_object()))
+        {
+            if (auto it = m_value.object->find(key); it != m_value.object->end())
+            {
+                return it->second;
+            }
+
+            return set_parent(m_value.object->operator[](json::object_t::key_type(key)));
```

Comment:
I updated the example that I'd suggested at https://github.com/nlohmann/json/issues/1529#issuecomment-808363396 to not copy at all, as std::map::{emplace, insert} use the comparator.

Would that example work here, to avoid the std::string copy, and the duplicate lookup via `operator[]`?

Reply:
Thanks1